calico/felix/test/test_fetcd_api.py

```diff
@@ -0,0 +1,185 @@
+            self.run_actor_loop()
+            self.assertTrue(self.etcd_api.client.write.called)
+
+    def test_update_felix_status_disabled(self):
+        """
+        Test reporting is disabled for reporting interval 0
+        """
+        self.finish_setup()
+        # We call gevent.sleep so that _status_reporting thread is executed -
+        # since no interval and ttl is given, _status_reporting thread returns
+        # instead of writing to etcd.
+        gevent.sleep(1)
+        self.assertFalse(self.etcd_api.write_to_etcd.called)
+
+    def test_update_felix_status_single(self):
+        """
+        Test felix status is updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=15,
+                          REPORTING_TTL_SECS=37)
+
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(1)
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+        self.etcd_api.write_to_etcd.assert_has_calls([status_call,
+                                                      uptime_call])
+
+    def test_update_felix_status_continuous(self):
+        """
+        Test felix status is being continuously updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=5,
+                          REPORTING_TTL_SECS=17)
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(5)
+
+        last_status_call = None
+        for update in range(5):
+            gevent.sleep(5)
+            # Check that both uptime and status were updated
+            self.etcd_api.write_to_etcd.assert_has_calls([status_call, uptime_call])
+            # Update last_status_call and reset write_to_etcd calls
+            last_status_call = self.etcd_api.write_to_etcd.mock_calls[0]
+            self.etcd_api.write_to_etcd.reset_mock()
+
+    @patch('calico.felix.fetcd._reconnect')
+    def test_update_felix_status_reconnects_on_etcd_exception(self, _reconnect):
+        """
+        Test felix status handles exceptions
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=4,
+                          REPORTING_TTL_SECS=12)
+        self.etcd_api.write_to_etcd = Mock(side_effect=EtcdException)
+
+        gevent.sleep(1)
+
+        self.assertTrue(_reconnect.called)
+
+
+class TestIfStatus(object):

```

Comment:
Thinking on these more, I think a good approach would be to define a generic class that converts a "validation" function into an object with an `__eq__`.  That'd be really useful as a general solution to this problem.  Right now, there's a lot of boilerplate to define one of these classes.
```
def is_valid_json(value):
    json.loads(value)  # Exception if not valid
    return True
c = call("foo", Check(is_valid_json), bar="baz")
```